apicast/src/oauth/keycloak.lua

```diff
@@ -0,0 +1,257 @@
+  end
+
+  ngx.log(ngx.DEBUG, '[rh-sso]: request: ', endpoint, ' response status: ', res.status, ' body: ', res.body)
+
+  if not key then
+    return nil, 'missing key'
+  end
+
+  return format_public_key(key)
+end
+
+function _M.load_configuration(client)
+  local endpoint = env.get('RHSSO_ENDPOINT')
+
+  if not endpoint then
+    return nil, 'missing endpoint configuration'
+  end
+
+  local http_client = http_ng.new{
+    backend = client,
+    options = {
+      ssl = { verify = false }
+    }
+  }
+
+  local keycloak_config = {
+    endpoint = endpoint,
+    authorize_url = resty_url.join(endpoint,'/protocol/openid-connect/auth'),
+    token_url = resty_url.join(endpoint,'/protocol/openid-connect/token'),
+    public_key = get_public_key(http_client, endpoint)
+  }
+  return keycloak_config
+end
+
+function _M.new(configuration)
+  return setmetatable({
+    config = configuration
+    }, mt)
+end
+
+function _M.respond_and_exit(status, body, headers)
+  -- TODO: is there a better way to populate the response headers?..
+  if headers then
+    for name,value in pairs(headers) do
+      ngx.header[name] = value
+    end
+  end
+
+  ngx.status = status
+  ngx.print(body)
+  ngx.exit(ngx.HTTP_OK)
+end
+
+function _M.respond_with_error(status, message)
+
+  --TODO: as per the RFC (https://tools.ietf.org/html/rfc6749#section-5.2) return WWW-Authenticate response header if 401
+  local headers = {
+    ['Content-Type'] = 'application/json;charset=UTF-8'
+  }
+  local err_msg = { error = message }
+  local body = cjson.encode(err_msg)
+  _M.respond_and_exit(status, body, headers)
+end
+
+function _M.authorize_check_params(params)
+  local response_type = params.response_type
+  local required_params = _M.params.response_type
+  if not response_type then return false, 'invalid_request' end
+  if not required_params[response_type] then return false, 'unsupported_response_type' end
+
+  for _,v in ipairs(required_params[response_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+
+  return true
+end
+
+function _M.token_check_params(params)
+  local grant_type = params.grant_type
+  local required_params = _M.params.grant_type
+  if not grant_type then return false, 'invalid_request' end
+  if not required_params[grant_type] then return false, 'unsupported_grant_type' end
+
+  for _,v in ipairs(required_params[grant_type]) do
+    if not params[v] then
+      return false, 'invalid_request'
+    end
+  end
+  return true
+end
+
+-- Parses the token - in this case we assume it's a JWT token
+-- Here we can extract authenticated user's claims or other information returned in the access_token
+-- or id_token by RH SSO
+local function parse_and_verify_token(self, jwt_token)
+  local jwt_obj = jwt:verify(self.config.public_key, jwt_token)
+
+  if not jwt_obj.verified then
+    ngx.log(ngx.INFO, "[jwt] failed verification for token, reason: ", jwt_obj.reason)
+    return jwt_obj, "JWT not verified"
```

Comment:
I'd change it to:
```
return jwt_obj, "JWT not verified: "..jwt_obj.reason
```
just not to lose the reason, in case we want to use it in `proxy` later.